tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -234,22 +253,27 @@ private Unsampled(String traceId) {
 
         @Override
         void fastStartSpan(String _operation, String parentSpanId, SpanType _type) {
-            startSpan(Optional.of(parentSpanId));
+            if (numberOfSpans == 0) {
+                originatingSpanId = Optional.of(parentSpanId);
+                topSpanId = Optional.of(Tracers.randomId());
+            }
+            numberOfSpans++;
         }
 
         @Override
         void fastStartSpan(String _operation, SpanType _type) {
+            if (numberOfSpans == 0) {
+                topSpanId = Optional.of(Tracers.randomId());
+            }
             numberOfSpans++;
         }
 
         @Override
         protected void push(OpenSpan span) {
-            startSpan(span.getParentSpanId());
-        }
-
-        private void startSpan(Optional<String> parentSpanId) {
             if (numberOfSpans == 0) {
-                originatingSpanId = parentSpanId;
+                // TODO: shouldn't this be span.getOriginatingSpanId?
+                originatingSpanId = span.getParentSpanId();
```

Comment:
When we push an existing span as the root span of a trace, shouldn't we preserve the originating span id?

Reply:
Yep i think you're correct.
However, because of the `if` above, this code would only ever be hit when the `span.getParentSpanId()` is the same as `span.getOriginatingSpanId()`, because `Trace.getOriginatingSpanId()` would always be empty when there are no spans.

Reply:
Honestly I don't even know why OpenSpan keeps a copy of `getOriginatingSpanId` because the Trace keeps a copy of it as well.
I think if we propagated this correctly in `Trace.deepCopy` then we could get rid of it from `OpenSpan` and this sort of question you're legitimately asking would become trivial to answer.

Reply:
Ah, I didn't realise `push` was only called from `startSpan` (in which case yeah we can assert that `span.getOriginatingSpanId()` is equal to `span.getParentId()` if we're empty)